### PR TITLE
chore: single workspace

### PR DESCRIPTION
**Pull request by Boshen**
## Summary

This PR puts `crates/node_binding` under the same workspace.

### <samp>🤖 Generated by Copilot at e93556b</samp>

This pull request experiments with a new workspace structure and a new enum for loading JavaScript code in `rspack`. It updates the test workflow, refactors and cleans up the `node_binding` and `rspack_binding_options` crates, and fixes a napi issue in the test environment. It also modifies the `Cargo.toml` files and the `.cargo/config.toml` file to support the new workspace and the new resolver.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

### <samp>🤖 Generated by Copilot at e93556b</samp>

*  Add a linker argument to fix a napi issue in the test environment ([link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-9a4f3e4537ebb7474452d131b0d969d89a51286f4269aac5ef268e712be17268L9-R17))
*  Modify the GitHub workflow for testing pull requests to only run on the `one-workspace` branch and to add more target architectures ([link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-553365cb26b1fc82e9fae16796d827431ccea4d72e0c1420cfed527de0e23fddL3-R24), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-553365cb26b1fc82e9fae16796d827431ccea4d72e0c1420cfed527de0e23fddR37-R44))
*  Comment out the exclusion of the `node_binding` and `rspack` crates from the workspace, and use the new resolver for feature unification ([link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542L3-R8))
*  Remove the `node-api` feature from the `rspack_binding_options` crate, and use a `JsLoaderRunner` enum instead of a `ThreadsafeFunction` struct for the loader runner parameter of the `apply` methods of the `RawOptions` and its substructures ([link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-dec41026e2dec6ab7a5bfe4f75643010bc788214bcd83257e4a00f5ec31338e7L15-R24), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-581cd5f0a115c33c2a3636a9284ba84c19ba6a4a89cea66a90e85a966e291a2aL75-R74), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-6112b39f2c1def8a0f89824413cfd6936f187aa6ca3712e1e6aeb4d34f67ae41L54-R54), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-6112b39f2c1def8a0f89824413cfd6936f187aa6ca3712e1e6aeb4d34f67ae41L97-R97), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-6112b39f2c1def8a0f89824413cfd6936f187aa6ca3712e1e6aeb4d34f67ae41L105-R110), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-6112b39f2c1def8a0f89824413cfd6936f187aa6ca3712e1e6aeb4d34f67ae41L134-R120), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-754529249fc831c71c3f399423ba1a92c432fe7c134e1fa51a6731831e0ab712L12), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-754529249fc831c71c3f399423ba1a92c432fe7c134e1fa51a6731831e0ab712L131-R130), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-b3d4e4741437a4af781bf0c7ffc91117b42dcefe17fd5ad19fb175e303124332L10), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-b3d4e4741437a4af781bf0c7ffc91117b42dcefe17fd5ad19fb175e303124332L136), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-3d952b60d975e6d9e1af6638d7a3cba2af369fb7f520a6cfc1e734e0b9b17ce4L2-R2), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-3d952b60d975e6d9e1af6638d7a3cba2af369fb7f520a6cfc1e734e0b9b17ce4L22-R43), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-3d952b60d975e6d9e1af6638d7a3cba2af369fb7f520a6cfc1e734e0b9b17ce4L91-R102), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-3d952b60d975e6d9e1af6638d7a3cba2af369fb7f520a6cfc1e734e0b9b17ce4L101), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-3d952b60d975e6d9e1af6638d7a3cba2af369fb7f520a6cfc1e734e0b9b17ce4L110), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-3d952b60d975e6d9e1af6638d7a3cba2af369fb7f520a6cfc1e734e0b9b17ce4L117), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-3d952b60d975e6d9e1af6638d7a3cba2af369fb7f520a6cfc1e734e0b9b17ce4L172), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-3d952b60d975e6d9e1af6638d7a3cba2af369fb7f520a6cfc1e734e0b9b17ce4L214), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-3d952b60d975e6d9e1af6638d7a3cba2af369fb7f520a6cfc1e734e0b9b17ce4L236), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-3d952b60d975e6d9e1af6638d7a3cba2af369fb7f520a6cfc1e734e0b9b17ce4L291), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-3d952b60d975e6d9e1af6638d7a3cba2af369fb7f520a6cfc1e734e0b9b17ce4L307), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-88d56d7c2fc780fbf03115c225e86e027a11aea2e0d1eb1dda2c0a1034aaaa59L6), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-88d56d7c2fc780fbf03115c225e86e027a11aea2e0d1eb1dda2c0a1034aaaa59L17), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-88d56d7c2fc780fbf03115c225e86e027a11aea2e0d1eb1dda2c0a1034aaaa59L170), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-88d56d7c2fc780fbf03115c225e86e027a11aea2e0d1eb1dda2c0a1034aaaa59L323-R320), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-88d56d7c2fc780fbf03115c225e86e027a11aea2e0d1eb1dda2c0a1034aaaa59L332), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-88d56d7c2fc780fbf03115c225e86e027a11aea2e0d1eb1dda2c0a1034aaaa59L355-R351), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-88d56d7c2fc780fbf03115c225e86e027a11aea2e0d1eb1dda2c0a1034aaaa59L417-R410), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-977596e157975dc4197eb665b26946ed33ed5fe3f497c0ddf25ec0a92558b062L9), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-977596e157975dc4197eb665b26946ed33ed5fe3f497c0ddf25ec0a92558b062L31-R30), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-ef5856a7fec8410383d58cc835932b263f5df692fd8d31fb1ea236182a162b00L9), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-ef5856a7fec8410383d58cc835932b263f5df692fd8d31fb1ea236182a162b00L152-R151), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-b4f30c9ffb52a6fa5ecf014abc68931db6c9ecfab8247c3280096aa69f416e62L3), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-b4f30c9ffb52a6fa5ecf014abc68931db6c9ecfab8247c3280096aa69f416e62L15-R14))
*  Add the `crate-type = ["lib", "cdylib"]` option to the `lib` section of the `rspack_binding_options` crate ([link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-581cd5f0a115c33c2a3636a9284ba84c19ba6a4a89cea66a90e85a966e291a2aL8-R9))
*  Remove the `profile.release` section from the `node_binding` crate ([link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-dec41026e2dec6ab7a5bfe4f75643010bc788214bcd83257e4a00f5ec31338e7L56-L62))
*  Comment out the dependency on `rspack` in the `dev-dependencies` section of the `rspack_error` crate ([link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-ecfe3e3156d8717f286e00893af607ada528d4b16217e90ab94e2cb090acb42dL21-R21))
*  Add the `&JsLoaderRunner::noop()` argument to the `apply` methods of the `raw` variables in the `build-cli.rs` example file and the `run_fixture.rs` file of the `rspack_testing` crate ([link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-e208790276fdcbb37cb458dbec3e9e7e2c09eed807529b0a3f7ae05ae9922193L12-R12), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-e208790276fdcbb37cb458dbec3e9e7e2c09eed807529b0a3f7ae05ae9922193L39-R41), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-a17804366fb795a1395f00e4406072e66850b85ad20d2c583480456734c517b4L4-R4), [link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-a17804366fb795a1395f00e4406072e66850b85ad20d2c583480456734c517b4L17-R19))
*  Delete the `crates/rspack_binding_options/build.rs` file ([link](https://github.com/web-infra-dev/rspack/pull/3035/files?diff=unified&w=0#diff-9685c0fa6a1beddf4a5def673436cb36cddddfa624ecc9d456a91ce0cc851137))

</details>

### CODE REVIEW

File: .github/workflows/test-pr.yml
Comment (Boshen) [inline .github/workflows/test-pr.yml:1]: Modified for testing, will revert.

**Review by hyf0 — DISMISSED** [review]

**Comment by Boshen** [comment]
Let's wait until Tuesday's release before rushing this into main.

File: scripts/check_rust_dependency.cjs
Comment (Boshen) [inline scripts/check_rust_dependency.cjs:1]: This no longer works with `resolver = 2`, and is unecessary anymore. cc @jerrykingxyz

File: crates/rspack_binding_options/Cargo.toml [hunk 1]
```diff
@@ -5,9 +5,8 @@ name       = "rspack_binding_options"
 repository = "https://github.com/web-infra-dev/rspack"
 version    = "0.1.0"
 
-[features]
-default  = ["napi-derive/noop"]
-node-api = ["dep:rspack_napi_shared"]
+[lib]
+crate-type = ["lib", "cdylib"]
```
[the remark this answers is not in the document]
Comment (Boshen) [reply crates/rspack_binding_options/Cargo.toml]: Removed.

File: Cargo.toml [hunk 2]
```diff
@@ -1,8 +1,6 @@
 [workspace]
-exclude = [
-  "crates/node_binding",
-] # Avoid including node binding, since feature unification will cause an linking issue. See: https://doc.rust-lang.org/cargo/reference/features.html#feature-unification
-members = ["crates/*"]
+members  = ["crates/*"]
+resolver = "2"
```
[the remark this answers is not in the document]
Comment (Boshen) [reply Cargo.toml]: Added comment for it, it's the default for `edition = 2021` anyway.

File: crates/rspack_binding_options/src/options/raw_module/js_loader.rs [hunk 3]
```diff
@@ -19,20 +17,30 @@ pub struct JsLoader {
   pub identifier: String,
 }
 
-#[cfg(feature = "node-api")]
 #[derive(Clone)]
-pub struct JsLoaderRunner(ThreadsafeFunction<JsLoaderContext, LoaderThreadsafeLoaderResult>);
+pub enum JsLoaderRunner {
+  ThreadsafeFunction(ThreadsafeFunction<JsLoaderContext, LoaderThreadsafeLoaderResult>),
+  Noop,
```
[the remark this answers is not in the document]
Comment (Boshen) [reply crates/rspack_binding_options/src/options/raw_module/js_loader.rs:25]: Done

**Review by hyf0 — APPROVED** [review]
you save my day

**Pull request merged**